src/Authentication/AuthenticationServiceFactory.cs

```diff
@@ -0,0 +1,61 @@
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Identity;
+using Microsoft.Graph.Cli.Utils;
+
+namespace Microsoft.Graph.Cli.Authentication;
+
+class AuthenticationServiceFactory {
+    public async Task<IAuthenticationService> GetAuthenticationServiceAsync(AuthenticationStrategy strategy, bool persistToken = false) {
+        switch (strategy) {
+            case AuthenticationStrategy.DeviceCode:
+                return await GetDeviceCodeAuthenticationServiceAsync(persistToken);
+            default:
+                throw new InvalidOperationException($"The authentication strategy {strategy} is not supported");
+        }
+
+    }
+
+    public async Task<TokenCredential> GetTokenCredentialAsync(AuthenticationStrategy strategy, bool persistToken = false) {
+        switch (strategy) {
+            case AuthenticationStrategy.DeviceCode:
+                return await GetDeviceCodeCredentialAsync(persistToken);
+            default:
+                throw new InvalidOperationException($"The authentication strategy {strategy} is not supported");
+        }
+    }
+
+    private async Task<DeviceCodeAuthenticationService> GetDeviceCodeAuthenticationServiceAsync(bool persistToken = false) {
+        var credential = await GetDeviceCodeCredentialAsync(persistToken);
+        return new(credential);
+    }
+
+    private async Task<DeviceCodeCredential> GetDeviceCodeCredentialAsync(bool persistToken) {
```

Comment:
I'd move all of that into the constructor of devicecodeauthenticationservice, so we don't end up with a bunch of construction code for each strategy

Reply:
This is an async function so moving it to a constructor would be tricky.